### PR TITLE
Add support for reading gzipped CSV files with read_csv when use_pyar…

### PR DESCRIPTION
**Pull request by ghuls**
…row=False.

Add support for reading gzipped CSV files with read_csv when use_pyarrow=False.
Gzipped files are decompressed with igzip of python-isal, if it is
installed as it is around 30% faster than the gzip module of python.
Pyarrows CSV parser is still much faster for reading gzipped CSV files
(at least twice as fast as igzip).

Timings:
  Reading gzipped CSV file with pyarrow:
    CPU times: user 45.4 s, sys: 15.2 s, total: 1min
    Wall time: 25.4 s

  Reading gzipped CSV file with igzip.open() and DataFrame.read_csv():
    CPU times: user 45.1 s, sys: 20.9 s, total: 1min 5s
    Wall time: 1min 5s

  Reading gzipped CSV file with gzip.open() and DataFrame.read_csv():
    CPU times: user 59.3 s, sys: 24.7 s, total: 1min 23s
    Wall time: 1min 24s

### CODE REVIEW

File: py-polars/polars/functions.py [hunk 1]
```diff
@@ -216,23 +216,51 @@ def read_csv(
...
-    )
+    if isinstance(file, str) and file.endswith(".gz"):
+        try:
+            # Try to use python-isal for faster gzip decompresssion.
+            from isal import igzip as gzip_mod
+        except ImportError:
+            # Use normal gzip module, if python-isal is not installed.
+            import gzip as gzip_mod
+
+        with gzip_mod.open(file, "rb") as fh:
```
Comment (ritchie46) [inline py-polars/polars/functions.py:227]: Maybe we can define a closure or a lambda above that takes only the new file (`fh` or `file`) as argument and captures the other args from the scope.


```python
def read_df(file):
    return DataFrame.read_csv(
	file=file,
	infer_schema_length=infer_schema_length,
	batch_size=batch_size,
	has_headers=has_headers,
	ignore_errors=ignore_errors,
	stop_after_n_rows=stop_after_n_rows,
	skip_rows=skip_rows,
	projection=projection,
	sep=sep,
	columns=columns,
	rechunk=rechunk,
	encoding=encoding,
	n_threads=n_threads,
	dtype=dtype,
	low_memory=low_memory,
        )


if ...:
     with gzip_mod.open(file, "rb") as fh:
          df = read_df(fh)
else:
    df = read_df(file)

```

**Review by ritchie46 — COMMENTED** [review]
Looks good. I have got one small suggestion.

File: py-polars/polars/functions.py (hunk 1 above)
Comment (ghuls) [reply py-polars/polars/functions.py:227]: So to make read_df a subfunction in read_csv?

Comment (ritchie46) [reply py-polars/polars/functions.py]: Yes, but then I am less likely to make mistakes when I add or remove some arguments.

**Comment by ritchie46** [comment]
:+1:

**Pull request merged**